src/test/regress/upgrade/custom_citus.py

```diff
@@ -0,0 +1,156 @@
+#!/usr/bin/env python3
+
+"""custom_citus
+Usage:
+    custom_citus --bindir=<bindir> --pgxsdir=<pgxsdir> --parallel=<parallel>
+
+Options:
+    --bindir=<bindir>              The PostgreSQL executable directory(ex: '~/.pgenv/pgsql-11.3/bin')
+    --pgxsdir=<pgxsdir>           	       Path to the PGXS directory(ex: ~/.pgenv/src/postgresql-11.3)
+    --parallel=<parallel>           how many configs to run in parallel
+"""
+
+import upgrade_common as common
+import threading
+import concurrent.futures
+from docopt import docopt
+import os, shutil
+import time
+import sys
+import inspect
+
+
+import config as cfg
+
+testResults = {}
+failCount = 0
+
+
+def run_for_config(config):
+    name = config.name
+    print("Running test for: {}".format(name))
+    start_time = time.time()
+    common.initialize_citus_cluster(
+        config.bindir, config.datadir, config.settings, config
+    )
+    if config.user == cfg.REGULAR_USER_NAME:
+        common.create_role(
+            config.bindir,
+            config.coordinator_port(),
+            config.node_name_to_ports.values(),
+            config.user,
+        )
+    copy_test_files(config)
+
+    exitCode = common.run_pg_regress_without_exit(
+        config.bindir,
+        config.pg_srcdir,
+        config.coordinator_port(),
+        cfg.CUSTOM_CREATE_SCHEDULE,
+        config.output_dir,
+        config.input_dir,
+        config.user,
+    )
+    common.save_regression_diff("create", config.output_dir)
+    if config.is_mx and config.worker_amount > 0:
+        exitCode |= common.run_pg_regress_without_exit(
+            config.bindir,
+            config.pg_srcdir,
+            config.random_worker_port(),
+            cfg.CUSTOM_SQL_SCHEDULE,
+            config.output_dir,
+            config.input_dir,
+            config.user,
+        )
+    else:
+        exitCode |= common.run_pg_regress_without_exit(
+            config.bindir,
+            config.pg_srcdir,
+            config.coordinator_port(),
+            cfg.CUSTOM_SQL_SCHEDULE,
+            config.output_dir,
+            config.input_dir,
+            config.user,
+        )
+
+    run_time = time.time() - start_time
+    testResults[name] = (
+        "SUCCESS"
+        if exitCode == 0
+        else "FAIL: see {}".format(config.output_dir + "/run.out")
+    )
```

Comment:
Inserting into a dictionary is threadsafe because of the Python GIL in this case. But I think it would be good not to rely on this. https://stackoverflow.com/a/55560667/2570866

Let's make this a return value of the function too.

Reply:
If we have such a concern, I feel like passing a lock would be better to do the sync in necessary parts as we can have more parts that needs syncing. I pushed what I meant, let me know what you think of: https://github.com/citusdata/citus/pull/5316/commits/3f603b65898c76c5a3eef4c87928a603f8103efc